ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,173 @@
+    }
+
+    event MessageDelivered(uint256 _nonce, bool _result);
+
+    function newParachainCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents,
+        uint256 parachainBlockNumber,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) public {
+        verifyCommitment(
+            commitment,
+            commitmentContents,
+            ourParachainMerkleProof,
+            parachainHeadsMMRProof
+        );
+        processCommitmentContents(commitmentContents);
+    }
+
+    function verifyCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) internal returns (bool success) {
+        // Prove we can get the MMRLeaf that is claimed to contain our Parachain Block Header
+        // BEEFYLightClient.verifyMMRLeaf(parachainHeadsMMRProof)
+        // BeefyLightClient{
+        // verifyMMRLeaf(parachainHeadsMMRProof) {
+        //MMRVerification.verifyInclusionProof(latestMMRRoot, parachainHeadsMMRProof)
+        // }
+        //}
+        //}
+        // returns mmrLeaf;
+
+        // Prove we can get the claimed parachain block header from the MMRLeaf
+        // allParachainHeadsMerkleTreeRoot = mmrLeaf.parachain_heads;
+        // MerkeTree.verify(allParachainHeadsMerkleTreeRoot, ourParachainMerkleProof)
+        // returns parachainBlockHeader
+
+        // Prove that the commitment is in fact in the parachain block header
+        // require(parachainBlockHeader.commitment == commitment)
+
+        // Validate that the commitment matches the commitment contents
+        require(
+            validateCommitment(commitment, commitmentContents),
+            "invalid commitment"
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >=
+                commitmentContents.messages.length * MAX_GAS_PER_MESSAGE,
+            "insufficient gas for delivery of all messages"
+        );
+
+        // Require all payloads are smaller than max_payload_size
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            require(
+                commitmentContents.messages[i].payload.length <=
+                    MAX_PAYLOAD_BYTE_SIZE,
+                "message payload bytesize exceeds maximum payload size"
+            );
+        }
+        return true;
+    }
+
+    function processCommitmentContents(
+        CommitmentContents memory commitmentContents
+    ) internal {
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            Message memory message = commitmentContents.messages[i];
+            require(message.nonce == lastProcessedNonce + 1, "invalid nonce");
```

Comment:
Why do we need message nonces? Since we process all messages from one commitments, a nonce per commitment would be sufficient.

Reply:
yeh, will create an issue for this potential optimization

Reply:
https://github.com/Snowfork/polkadot-ethereum/issues/251